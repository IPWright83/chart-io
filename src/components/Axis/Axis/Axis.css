.g-axis .axis {
    user-select: none;
}
