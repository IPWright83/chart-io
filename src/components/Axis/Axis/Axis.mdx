import { Meta, Story, Canvas } from "@storybook/addon-docs/blocks";

# Axis Component

This `<Axis>` component allows you to define an Axis on a chart. Underneath this automatically creates a `d3.Scale` too, which is used by most plots to correctly translate data to on screen co-ordinates. The component takes an array of fields that you wish to display along that axis, for example:

```
<YAxis fields={["Profit", "Loss"]} />
```

Typically you should use the `<XAxis>` or `<YAxis>` component depending on the orientation, unless you want more fine grained control in which case you can use an `<Axis>`. Each of these components will attempt to automatically determine the best axis & scale type for the data.

### Props

| Prop            | Type            | Default | Note                                                                                                                                                                      |
| --------------- | --------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `fields`        | `Array<string>` |         | The set of fields that this axis represents. These should correspond to the fields being displayed on this Axis within the data.                                          |
| `position`      | `string`        |         | The position of the axis. One of `top`, `bottom`, `left` or `right`. This will use the appropriate [d3.axis](https://github.com/d3/d3-axis#axisTop) orientation function. |
| `showGridlines` | `boolean`       | true    | Determines whether gridlines should be shown against this axis.                                                                                                           |
| `tickFormat`    | `function`      | null    | See https://github.com/d3/d3-axis#axis_tickFormat                                                                                                                         |
| `tickSizeInner` | `number`        | 6       | See https://github.com/d3/d3-axis#axis_tickSizeInner.                                                                                                                     |
| `tickSizeOuter` | `number`        | 6       | See https://github.com/d3/d3-axis#axis_tickSizeOuter.                                                                                                                     |
| `tickPadding`   | `number`        | 3       | See https://github.com/d3/d3-axis#axis_tickPadding.                                                                                                                       |
| `title`         | `string`        | null    | The title of the axis. This will default to the value in field(s)                                                                                                         |

<br />

## Discrete Scales

Discrete scales are best suited for `<Bar>` or `<Column>` plots but can also be used for `<Scatter>` plots too and are the usual scale for the following types and will produce axis like the following examples:

-   String
-   Boolean

<Canvas>
    <Story id="components-axis-types--string-axis" />
</Canvas>

<Canvas>
    <Story id="components-axis-types--boolean-axis" />
</Canvas>

## Numerical Axis

Numerical continuous scales are best for plotting `<Area>`, `<Line>` and `<Scatter` plots, although are freqeuently the scale of choice for the `<YAxis>` on a `<Column>` chart for example. Here are some examples showing different data types and ranges:

-   Integer
-   Double

<Canvas>
    <Story id="components-axis-types--integer-axis" />
</Canvas>

<Canvas>
    <Story id="components-axis-types--double-axis" />
</Canvas>

<Canvas>
    <Story id="components-axis-types--double-axis-only-negative-axis" />
</Canvas>

<Canvas>
    <Story id="components-axis-types--double-axis-mixed-sign-axis" />
</Canvas>

### Aggregate

Sometimes when using multiple fields, rather than sharing an Axis, you instead want to aggregate those values together. This is often the case if you're stacking your data. To do this you can just set the `aggregate` property to true. If you omit this value then plots will attempt to warn you if data is going to be cropped.

```
<YAxis fields={["Sales_Revenue", "Consulting_Revenue"]} aggregate />
```

### ScaleType

There are occasions where you may want to speicfy the type of d3.Scale that is used underneath. To do this you can specify the `scaleType` explicitly. Currently you can choose one of the following options:

-   power
-   linear
-   time
-   log
-   band
-   point

A common example when you may wish to do this is when if you wish to display numerical data on a `<Column>` or `<Bar>` plot. These plots require a **band** scale but a linear scale would be returned by default.

```
<XAxis fields={["dayOfWeek"]} scaleType="band" />
```

## Time Axis

Time scales are best for plotting `<Area>`, `<Line>` and showing data across time. The data types supported are:

-   Date
-   DateTime

<Canvas>
    <Story id="components-axis-types--date-axis" />
</Canvas>

<Canvas>
    <Story id="components-axis-types--date-time-axis" />
</Canvas>

## Customization

There are a number of customizations that can be used to customize the apperance of an axis. One of the most common is the insertion of gridlines:

<Canvas>
    <Story id="components-axis-customization--gridlines" />
</Canvas>

However you can also adjust:

-   `tickPadding`
-   `tickSizeOuter`
-   `tickSizeInner`
-   `title`

It is also relatively easy to build a custom Axis that hooks directly into the chart. Note that if you're not using an `XAxis` or `YAxis` you may need to manually define a `<Scale>` for your chart.

<Canvas>
    <Story id="components-axis-customization--fully-custom-axis" />
</Canvas>

The above can be produced with:

```
const CustomTimeAxis = ({ fields }) => {
    const axis = useRef(null);

    const field = fields[0];
    const scale = useSelector((s) => chartSelectors.scales.getScale(s, field));

    useEffect(() => {
        if (axis.current && scale) {
            const selection = d3.select(axis.current);

            const timeFormat = d3.timeFormat("%H:%M");
            const d3Axis = d3
                .axisBottom(scale)
                .ticks(d3.timeHour.every(1))
                .tickSizeInner(30)
                .tickSizeOuter(30)
                .tickFormat((value, index) => (index % 3 === 0 ? timeFormat(value) : null));

            d3Axis.scale(scale);

            selection.call(d3Axis);
        }
    }, [scale]);

    return <g className="customAxis" ref={axis} />;
};

const CustomAxisTemplate = (args) => (
    <Chart {...args} data={axisData}>
        <XScale fields={args.fields} scaleType="time" />
        <CustomTimeAxis fields={args.fields} />
    </Chart>
);
```
