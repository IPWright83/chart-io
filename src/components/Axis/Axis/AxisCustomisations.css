.customAxis .tick:nth-child(3n) line,
.customAxis .tick:nth-child(3n + 1) line {
  stroke-dasharray: 30;
  stroke-dashoffset: 22;
}

.customAxis .tick text {
  transform: translate(22px, -12px);
}
