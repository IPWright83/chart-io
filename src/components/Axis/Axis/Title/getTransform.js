/**
 * Obtain the transform for the Axis title
 * @param  {String} position     The position of the axis [left, right, top, bottom]
 * @param  {Number} width        The width for the chart
 * @param  {Number} height       The height for the chart
 * @param  {Number} margin       The margin object for the chart
 * @return {String}              The transform
 */
const getTransform = (position, width, height, margin) => {
    if (width === 0 || height === 0) {
        return "translate(0, 0)";
    }

    const actualWidth = width - margin.left - margin.right;
    const actualHeight = height - margin.top - margin.bottom;
    const halfWidth = actualWidth / 2;
    const halfHeight = actualHeight / 2;

    switch (position) {
        case "left":
            return `translate(${margin.left / 2 - 25}, ${halfHeight + margin.top / 2})rotate(270)`;
        case "right":
            return `translate(${width - margin.right / 2 + 25}, ${halfHeight + margin.top / 2})rotate(270)`;
        case "top":
            return `translate(${halfWidth + margin.left / 2}, ${margin.top - 25})`;
        case "bottom":
            return `translate(${halfWidth + margin.left / 2}, ${height - margin.bottom + 35})`;
        default:
            throw new Error(`Invalid position: ${position}`);
    }
};

export { getTransform };
