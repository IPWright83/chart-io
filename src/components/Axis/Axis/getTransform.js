/**
 * Obtain the transform for the Axis group
 * @param  {String} position     The position of the axis [left, right, top, bottom]
 * @param  {Number} width        The width for the chart
 * @param  {Number} height       The height for the chart
 * @param  {Number} margin       The margin object for the chart
 * @return {String}              The transform
 */
const getTransform = (position, width, height, margin) => {
    if (width === 0 || height === 0) {
        return "translate(0, 0)";
    }

    switch (position) {
        case "left":
            return `translate(${margin.left}, 0)`;
        case "right":
            return `translate(${width - margin.right}, 0)`;
        case "top":
            return `translate(0, ${margin.top})`;
        case "bottom":
            return `translate(0, ${height - margin.bottom})`;
        default:
            throw new Error(`Invalid position: ${position}`);
    }
};

export { getTransform };
