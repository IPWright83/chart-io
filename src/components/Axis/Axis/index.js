import { withSVG } from "../../../hoc";
import { Axis as AxisBase } from "./Axis";

const Axis = withSVG(AxisBase, "axis");

export { Axis };
