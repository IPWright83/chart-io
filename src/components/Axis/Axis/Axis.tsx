import "./Axis.css";

import * as d3 from "d3";

import PropTypes from "prop-types";
import React, { useEffect, useRef } from "react";
import { useSelector } from "react-redux";

import { getD3Axis } from "./getD3Axis";
import { getTransform } from "./getTransform";
import { Gridlines } from "./Gridlines";
import { Title } from "./Title";

import { IPosition } from "../../../types";

import { chartSelectors } from "../../../store";

export interface IAxis {
    position: IPosition;
    fields: string[];
    tickSizeInner: number;
    tickSizeOuter: number;
    tickPadding: number;
    showGridlines: boolean;
    title?: string;
    tickFormat?: Function;
    ticks: any[];
}

/**
 * Represents an Axis component
 * @return The Axis component
 */
const Axis = ({
    position,
    fields,
    tickSizeInner = 6,
    tickSizeOuter = 6,
    tickPadding = 3,
    ticks,
    showGridlines = true,
    title,
    tickFormat,
}: IAxis) => {
    if (fields.length === 0) {
        throw new Error(
            "Unable to render an Axis without a field. Ensure that you have provided at least one field in the 'fields' prop.",
        );
    }

    const field = fields[0];
    const width = useSelector((s) => chartSelectors.dimensions.width(s));
    const height = useSelector((s) => chartSelectors.dimensions.height(s));
    const margin = useSelector((s) => chartSelectors.dimensions.margin(s));
    const scale = useSelector((s) => chartSelectors.scales.getScale(s, field));
    const animationDuration = useSelector((s) => chartSelectors.animationDuration(s));
    const transform = getTransform(position, width, height, margin);

    // React will own the axis containers, but D3 will own the axis themselves
    const axis = useRef(null);

    // Render the x-axis using D3
    useEffect(() => {
        if (axis.current && scale) {
            const selection = d3.select(axis.current).transition().duration(animationDuration);

            // Create the D3 axis renderer
            const d3Axis = getD3Axis(position, scale);

            // Set some scale props
            d3Axis
                .tickSizeInner(tickSizeInner)
                .tickSizeOuter(tickSizeOuter)
                .tickPadding(tickPadding)
                .tickFormat(tickFormat)
                .ticks(ticks);

            // Render the axis
            selection.call(d3Axis);
        }
    }, [position, axis, scale, animationDuration, tickPadding, tickSizeInner, tickSizeOuter, showGridlines]);

    return (
        <React.Fragment>
            <Title position={position} title={title} fields={fields} />
            <g transform={transform}>
                {showGridlines ? <Gridlines position={position} scale={scale} ticks={ticks} /> : null}
                <g className={`axis-${position}`} ref={axis} />
            </g>
        </React.Fragment>
    );
};

export { Axis };
