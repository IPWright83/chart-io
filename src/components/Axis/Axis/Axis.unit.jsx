import * as d3 from "d3";
import React from "react";
import { Provider } from "react-redux";

import { render } from "@testing-library/react";

import { getTransform } from "./getTransform";
import { Axis } from "./Axis";

describe("Axis", () => {
    const width = 1000;
    const height = 500;
    const margin = { left: 10, right: 20, top: 30, bottom: 40 };

    const store = {
        getState: () => ({
            chart: {
                dimensions: {
                    width,
                    height,
                    margin,
                },
                scales: {
                    x: d3.scaleLinear().domain([0, width]).range([0, 100]),
                    y: d3.scaleLinear().domain([0, width]).range([0, 50]),
                },
            },
        }),
        dispatch: () => {},
        subscribe: () => {},
    };

    describe("component", () => {
        it("renders a left axis", async () => {
            const layer = { current: document.createElement("custom") };

            const { asFragment } = render(
                <Provider store={store}>
                    <Axis layer={layer} title="y" position="left" fields={["y"]} showGridlines={false} />
                </Provider>,
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it("renders a right axis", () => {
            const layer = { current: document.createElement("custom") };

            const { asFragment } = render(
                <Provider store={store}>
                    <Axis layer={layer} title="y" position="right" fields={["y"]} showGridlines={false} />
                </Provider>,
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it("renders a top axis", () => {
            const layer = { current: document.createElement("custom") };

            const { asFragment } = render(
                <Provider store={store}>
                    <Axis layer={layer} title="x" position="top" fields={["x"]} showGridlines={false} />
                </Provider>,
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it("renders a bottom axis", () => {
            const layer = { current: document.createElement("custom") };

            const { asFragment } = render(
                <Provider store={store}>
                    <Axis layer={layer} title="x" position="left" fields={["x"]} showGridlines={false} />
                </Provider>,
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });

    describe("getTransform", () => {
        describe("should return correct transform", () => {
            it("for left axis", () => {
                expect(getTransform("left", width, height, margin)).toBe("translate(10, 0)");
            });
            it("for right axis", () => {
                expect(getTransform("right", width, height, margin)).toBe("translate(980, 0)");
            });
        });

        describe("should return correct transform", () => {
            it("for top axis", () => {
                expect(getTransform("top", width, height, margin)).toBe("translate(0, 30)");
            });
            it("for bottom axis", () => {
                expect(getTransform("bottom", width, height, margin)).toBe("translate(0, 460)");
            });
        });

        describe("should return no translation", () => {
            it("for a 0 width", () => {
                expect(getTransform("top", 0, height, margin)).toBe("translate(0, 0)");
            });
            it("for a 0 height", () => {
                expect(getTransform("top", width, 0, margin)).toBe("translate(0, 0)");
            });
        });

        it("throws an error with an invalid position", () => {
            expect(() => getTransform("invalid", width, height, margin)).toThrow();
        });
    });
});
