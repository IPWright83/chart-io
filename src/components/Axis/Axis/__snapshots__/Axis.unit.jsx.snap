// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Axis component renders a bottom axis 1`] = `
<DocumentFragment>
  <svg>
    <text
      class="chart-it axis-title"
      transform="translate(-20, 230)rotate(270)"
    >
      x
    </text>
    <g
      transform="translate(10, 0)"
    >
      <g
        class="chart-it axis axis-left"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="color: rgb(51, 51, 51); stroke-opacity: 0.8; stroke-width: 1;"
        text-anchor="end"
      >
        <path
          class="domain"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,0.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,10.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,20.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,30.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,40.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,50.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,60.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,70.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,80.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,90.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,100.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Axis component renders a left axis 1`] = `
<DocumentFragment>
  <svg>
    <text
      class="chart-it axis-title"
      transform="translate(-20, 230)rotate(270)"
    >
      y
    </text>
    <g
      transform="translate(10, 0)"
    >
      <g
        class="chart-it axis axis-left"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="color: rgb(51, 51, 51); stroke-opacity: 0.8; stroke-width: 1;"
        text-anchor="end"
      >
        <path
          class="domain"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,0.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,5.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,10.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,15.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,20.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,25.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,30.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,35.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,40.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,45.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,50.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          />
        </g>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Axis component renders a right axis 1`] = `
<DocumentFragment>
  <svg>
    <text
      class="chart-it axis-title"
      transform="translate(1015, 230)rotate(270)"
    >
      y
    </text>
    <g
      transform="translate(980, 0)"
    >
      <g
        class="chart-it axis axis-right"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="color: rgb(51, 51, 51); stroke-opacity: 0.8; stroke-width: 1;"
        text-anchor="start"
      >
        <path
          class="domain"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,0.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,5.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,10.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,15.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,20.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,25.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,30.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,35.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,40.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,45.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0,50.5)"
        >
          <line
            stroke="currentColor"
            x2="6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="9"
          />
        </g>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Axis component renders a top axis 1`] = `
<DocumentFragment>
  <svg>
    <text
      class="chart-it axis-title"
      transform="translate(490, 5)"
    >
      x
    </text>
    <g
      transform="translate(0, 30)"
    >
      <g
        class="chart-it axis axis-top"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="color: rgb(51, 51, 51); stroke-opacity: 0.8; stroke-width: 1;"
        text-anchor="middle"
      >
        <path
          class="domain"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(0.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(10.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(20.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(30.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(40.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(50.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(60.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(70.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(80.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(90.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
        <g
          class="tick"
          opacity="0.000001"
          transform="translate(100.5,0)"
        >
          <line
            stroke="currentColor"
            y2="-6"
          />
          <text
            dy="0em"
            fill="currentColor"
            y="-9"
          />
        </g>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
